### PR TITLE
[AppService] Fix #15761: az webapp config ssl sync: Add command to sync imported KV certificate on webapp.

### PR DESCRIPTION
**Pull request by singlessis**
Description
This PR is adding a command to az webapp config ssl that allows customer to sync an imported KV cert. It effectively makes a create call, and updates the certificate in the app service database with whatever is in the KV.

Testing Guide
az webapp config ssl sync --resource-group testWestEurope --name SofiTestWestEurope --certificate-name testWestEurope-keyvaultSofiTest-test2048 --key-vault testKV

This certificate, if it is already added to that webapp, will be synced with whatever is in KV. If there is a new cert in KV, we will update the version in our system.

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AppService

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 1]
```diff
@@ -229,6 +229,9 @@ def load_arguments(self, _):
         with self.argument_context(scope + ' config ssl import') as c:
             c.argument('key_vault', help='The name or resource ID of the Key Vault')
             c.argument('key_vault_certificate_name', help='The name of the certificate in Key Vault')
+        with self.argument_context(scope + ' config ssl sync') as c:
+            c.argument('certificate_name', help='The name of the certificate to be synced')
+            c.argument('key_vault', help='The keyvault backing this cert, where the data will be synced from')
```
Comment (qwordy) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:234]: What is the supported format? Name? ID?

Comment (singlessis) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py]: KeyVault can be both name and ID, I added some notes to clarify that

**Review by qwordy — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 2]
```diff
@@ -353,7 +353,18 @@
 short-summary: Create a Managed Certificate for a hostname in a function app.
 examples:
   - name: Create a Managed Certificate for $fqdn.
-    text: az functionapp config ssl create --resource-group MyResourceGroup --name MyWebapp --hostname $fqdn
+    text: az functionapp config ssl create --resource-group MyResourceGroup --name MyFunctionApp --hostname $fqdn
+"""
+
+helps['functionapp config ssl sync'] = """
+type: command
+short-summary: Sync an imported certificate on a functionapp with its object in KV.
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:361]: try not to use acronyms like KV - use full name "KeyVault"

Comment (panchagnula) [reply src/azure-cli/azure/cli/command_modules/appservice/_help.py:361]: this is not syncing cert - this is syncing the webapp to the latest version of the certificate from KeyVault; that is associated with the webapp/function name. Please update the summary to reflect this.

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 3]
```diff
@@ -353,7 +353,18 @@
...
+examples:
+  - name: Sync a certificate with the KV object, using KeyVault name.
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:363]: Same comment we are not syncing cert - cert already exists in the AKV.

File: src/azure-cli/azure/cli/command_modules/appservice/_help.py [hunk 4]
```diff
@@ -1336,6 +1347,14 @@
     text: az webapp config ssl create --resource-group MyResourceGroup --name MyWebapp --hostname $fqdn
 """
 
+helps['webapp config ssl sync'] = """
+type: command
+short-summary: Sync an imported certificate on a webapp with its object in KV.
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/_help.py:1352]: Same here. Check with Yutang incase you need help.

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 5]
```diff
@@ -229,6 +229,9 @@ def load_arguments(self, _):
         with self.argument_context(scope + ' config ssl import') as c:
             c.argument('key_vault', help='The name or resource ID of the Key Vault')
             c.argument('key_vault_certificate_name', help='The name of the certificate in Key Vault')
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:231]: Shouldn't this be marked as deprecated in future releases? Moving forward only single parameter should be supported.

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 6]
```diff
@@ -229,6 +229,9 @@ def load_arguments(self, _):
...
+        with self.argument_context(scope + ' config ssl sync') as c:
+            c.argument('certificate_name', help='The name of the certificate to be synced')
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:233]: this can be certificate name or certificate resource ID?

Comment (panchagnula) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py:233]: Actually why do you need the certName? can't we get the latest version of the cert from the AKV & just use that?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 7]
```diff
@@ -2445,6 +2445,44 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
                                                 certificate_envelope=kv_cert_def)
 
 
+def sync_kv_cert(cmd, resource_group_name, name, certificate_name, key_vault):
+    Certificate = cmd.get_models('Certificate')
+    client = web_client_factory(cmd.cli_ctx)
+    webapp = client.web_apps.get(resource_group_name, name)
+    if not webapp:
+        err_msg = "'{}' app doesn't exist in resource group {}".format(name, resource_group_name)
+        raise ResourceNotFoundError(err_msg)
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2454]: Why do we need this specific check? Is this something we do on other commands? We should have a common validator that we use on other commands already that should check this.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 8]
```diff
@@ -2445,6 +2445,44 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+    server_farm_id = webapp.server_farm_id
+    location = webapp.location
+
+    cert = client.certificates.get(resource_group_name, certificate_name, filter=f"ServerFarmId eq {'server_farm_id'}")
+
+    if cert is None:
+        no_cert_msg = "Cert with name {} is not associated with this site".format(certificate_name)
+        logger.warning(no_cert_msg)
+        return
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2463]: We currently only allow syncing to the latest version of the cert on AKV using the AKV apis to get the latest version Ex. https://docs.microsoft.com/en-us/rest/api/keyvault/getcertificates/getcertificates - so don't think we need a cert name at all here.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -2445,6 +2445,44 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+    if cert.key_vault_id is None:
+        no_kv_msg = 'In order to sync a cert, it must be associated with a key vault'
+        logger.warning(no_kv_msg)
+        return
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2467]: we add basic validations need under validators.py

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -2445,6 +2445,44 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+    if is_valid_resource_id(key_vault):
+        if key_vault != cert.key_vault_id:
+            kv_not_match = "Key Vault id does not match the key vault associated with this cert"
+            logger.warning(kv_not_match)
+            return
+    else:
+        if parse_resource_id(cert.key_vault_id)['name'] != key_vault:
+            kv_name_not_match = "Key Vault name does not match the key vault name associated with this cert"
+            logger.warning(kv_name_not_match)
+            return
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2477]: if this is needed - should be on validators.py

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 11]
```diff
@@ -2445,6 +2445,44 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+
+    kv_cert_def = Certificate(location=location, key_vault_id=cert.key_vault_id, password='',
+                              key_vault_secret_name=cert.key_vault_secret_name, server_farm_id=server_farm_id)
+
+    return client.certificates.create_or_update(name=certificate_name, resource_group_name=resource_group_name,
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2482]: Does this preserve the binding on the app as well if exists? on ARM template we pass the binging info as well, can we make sure to add this to the test case.

File: src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py [hunk 12]
```diff
@@ -1418,6 +1418,50 @@ def test_webapp_ssl_import_crossrg(self, resource_group, kv_resource_group):
...
+        # https://docs.microsoft.com/azure/app-service-web/web-sites-configure-ssl-certificate#bkmk_ssopenssl
+        pfx_file = os.path.join(TEST_DIR, 'server.pfx')
+        cert_password = 'test'
+        cert_name = 'test-cert'
+
+        pfx_file_2 = os.path.join(TEST_DIR, 'server2.pfx')
+        cert_password_2 = 'test2'
+        cert_thumbprint_2 = '4E659097C7F3C261847909BD596910C05541215C'
+
+        # we configure tags here in a hope to capture a repro for https://github.com/Azure/azure-cli/issues/6929
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py:1437]: what is this comment about??

File: src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py [hunk 13]
```diff
@@ -1418,6 +1418,50 @@ def test_webapp_ssl_import_crossrg(self, resource_group, kv_resource_group):
...
+            'webapp create -g {} -n {} --plan {}'.format(resource_group, webapp_name, plan_name))
+        self.cmd('keyvault create -g {} -n {}'.format(resource_group, kv_name))
+        self.cmd('keyvault set-policy -g {} --name {} --spn {} --secret-permissions get'.format(
+            resource_group, kv_name, 'Microsoft.Azure.WebSites'))
+        self.cmd('keyvault certificate import --name {} --vault-name {} --file "{}" --password {}'.format(
+            cert_name, kv_name, pfx_file, cert_password))
+
+        cert_import_name = self.cmd('webapp config ssl import --resource-group {} --name {}  --key-vault {} --key-vault-certificate-name {}'.format(resource_group, webapp_name, kv_name, cert_name)).get_output_in_json()['name']
+
+        self.cmd('keyvault certificate import --name {} --vault-name {} --file "{}" --password {}'.format(
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py:1450]: Can we add a validation here to check that the latest version of the cert on AKV resource is in-fact what we are expecting before we proceed with sync?

File: src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py [hunk 14]
```diff
@@ -1418,6 +1418,50 @@ def test_webapp_ssl_import_crossrg(self, resource_group, kv_resource_group):
...
+        self.cmd('webapp config ssl sync --resource-group {} --name {}  --key-vault {} --certificate-name {}'.format(resource_group, webapp_name, kv_name, cert_import_name), checks=[
+            JMESPathCheck('thumbprint', cert_thumbprint_2)
+        ])
+
+        self.cmd('webapp config ssl bind -g {} -n {} --certificate-thumbprint {} --ssl-type {}'.format(resource_group, webapp_name, cert_thumbprint_2, 'SNI'), checks=[
+            JMESPathCheck("hostNameSslStates|[?name=='{}.azurewebsites.net']|[0].sslState".format(
+                webapp_name), 'SniEnabled'),
+            JMESPathCheck("hostNameSslStates|[?name=='{}.azurewebsites.net']|[0].thumbprint".format(
+                webapp_name), cert_thumbprint_2)
+        ])
```
Comment (panchagnula) [inline src/azure-cli/azure/cli/command_modules/appservice/tests/latest/test_webapp_commands.py:1462]: why are we adding binding post sync? Sync should preserve  the existing binding. I would update the test to include binding on the app before calling sync.

**Review by panchagnula — CHANGES_REQUESTED** [review]
Please check comments & get the build to pass.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 7 above)
Comment (singlessis) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:2454]: This was already done in the import certificate command, but I can remove it from both.

**Comment by qwordy** [comment]
Hi @singlessis, could you resolve Sisira's comments?

**Comment by singlessis** [comment]
@qwordy the checks have been stuck for a few days, without success or failure, is there any way to get more information on why? thanks!

**Comment by qwordy** [comment]
> 
> 
> @qwordy the checks have been stuck for a few days, without success or failure, is there any way to get more information on why? thanks!

@singlessis The CI system in Github is not stable. I have manually triggered it by sending `/azp run`.

**Comment by panchagnula** [comment]
@calvinsID can you review this PR please?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 15]
```diff
@@ -2476,6 +2476,74 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+        for kv in key_vaults:
+            if key_vault.lower() == kv.name.lower():
+                key_vault_id = kv.id
+                break
+    else:
+        key_vault_id = key_vault
+
+    if thumbprint:
+        certs = list(client.certificates.list_by_resource_group(resource_group_name,
+                                                                filter=f"ServerFarmId eq {'server_farm_id'}"))
```
Comment (calvinsID) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2502]: I think the single quotes should be outside of the `{}`. Also, just double checking but `filter` won't be used until the SDK/swagger is updated right?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 16]
```diff
@@ -2476,6 +2476,74 @@ def import_ssl_cert(cmd, resource_group_name, name, key_vault, key_vault_certifi
...
+            return
+
+        kv_cert_def = Certificate(location=location, key_vault_id=cert.key_vault_id, password='',
+                                  key_vault_secret_name=cert.key_vault_secret_name, server_farm_id=server_farm_id)
+
+        return client.certificates.create_or_update(name=cert.name, resource_group_name=resource_group_name,
+                                                    certificate_envelope=kv_cert_def)
+
+    certs = list(client.certificates.list_by_resource_group(resource_group_name,
+                                                            filter=f"ServerFarmId eq {'server_farm_id'}"))
```
Comment (calvinsID) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:2520]: same as above

**Review by calvinsID — APPROVED** [review]

**Review by calvinsID — APPROVED** [review]
Approved after resolving comments

**Pull request closed without merging**